File: msf.py
import sys
import heapq
class Arco: 
	def __init__(self, u:int, v:int, w:int):
		self.u = u
		self.v = v
		self.w = w
		self.msf = False
	def __lt__(self, altro):
		if not isinstance(altro, Arco):
			return NotImplemented
		return self.w < altro.w
	def __eq__(self, altro):
		return (self.u == altro.u and self.v == altro.v and self.w == altro.w)

nNodi:int = -1 #Numero di nodi nel grafo
nArchi:int = 0 #Numero di archi nel grafo
costoMSF:int = 0 #Costo della MSF
numCoCo:int #Numero di componenti connesse
hash_map:dict = {} #Tabella hash per gli archi 

if(len(sys.argv) < 3): #Controlla che siano stati passati i giusti argomenti
	sys.sdterr.write("Utilizzo: msf.py file_grafo file_archi")
	sys.exit(1)

#File per grafo e archi
file_grafo:str = sys.argv[1]
file_operazioni:str = sys.argv[2] 
archi:list[list[Arco]] = [] #Lista di archi per ricerca msf

#-------------------------------------LETTURA FILE GRAFO--------------------------------------------------
with open(file_grafo, "r") as file:
	for linea in file: 
		campi = linea.strip().split()
		if(not campi):
			continue
		if(campi[0] == "c"): #Ignora i commenti e continua
			continue
		elif(campi[0] == "p"): #Configura il numero di nodi e di archi
			nNodi = int(campi[2]) +1
			archi = [[] for _ in range(nNodi +1)]
			nArchi = int(campi[3])
		elif(campi[0] == "a"): #Inizializza l'arco e i suoi campi e inserisce nella hashtable
			if (nNodi == -1): #Se non è stato ancora letto il numero di nodi, segnala errore
				sys.stderr.write("Errore: formato file grafo non valido")
				sys.exit(1)
			u:int = int(campi[1])
			v:int = int(campi[2])
			w:int = int(campi[3])
			a:Arco = Arco(u, v, w)
			key:int = hash((min(u,v),max(u,v))) #Genera chiave normalizzando gli indici 
			hash_map[key] = a
			archi[a.u].append(a) #Aggiunge l'arco alla lista degli archi uscenti dal nodo u
			rev:Arco = Arco(v, u, w)
			archi[a.v].append(rev) #Aggiunge l'arco alla lista degli archi uscenti dal nodo v

numCoCo = nNodi #Inizializza il numero di componenti connesse al numero di nodi 

#-------------------------------------LETTURA OPERAZIONI---------------------------------------------------
with open(file_operazioni) as file:
	for linea in file:
		campi = linea.strip().split()
		if(not campi):
			continue
		if(campi[0] == "c"):
			continue
		elif(campi[0] == "+"):
			u:int = int(campi[1])
			v:int = int(campi[2])
			w:int = int(campi[3])
			a:Arco = Arco(u,v,w)
			key = hash((min(u,v),max(u,v)))
			if(key in hash_map):
				sys.stderr.write("Inserimento di arco duplicato")
				continue
			hash_map[key] = a
			archi[a.u].append(a)
			rev:Arco = Arco(v, u, w)
			archi[a.v].append(rev)
			nArchi+=1
		elif(campi[0] == "-"):
			u:int = int(campi[1])
			v:int = int(campi[2])
			key = hash((min(u,v),max(u,v))) #Calcola la chiave di hash
			#Se l'arco non esiste avvisa su stderr
			if(key not in hash_map):
				sys.stderr.write("Tentata rimozione di arco non esistente"+ campi[1] + campi[2])
				continue
			#Altrimenti elimina l'arco sia dalla hashmap che dalla lista per msf
			a:Arco = hash_map.pop(key)
			
			archi[a.u].remove(a)
			rev:Arco = Arco(v, u, a.w)
			
			archi[a.v].remove(rev)
			nArchi-=1
#-------------------------------------PRIM---------------------------------------------------
contaVisite:int = 0 #Numero di nodi visitati
visitati:list[bool] = [False] * (nNodi+1) #bitmap dei nodi visitati
msf:list[int] = [] #Nodi nel mst
heap_archi:list[Arco] = [] #Heap degli archi 
sorgente:int

def visita(v): #Dato un nodo, lo segna come visitato e aggiunge gli archi da esso uscenti allo heap
	visitati[v] = True
	for a in archi[v]:
		if(not visitati[a.v]):
			heapq.heappush(heap_archi, a) #Inserisce l'arco nel min heap mantenendo valida la proprietà di heap
		

while(contaVisite < nNodi): #Finchè non sono stativi sitati tutti i nodi
	sorgente = 0 
	#Cerca la sorgente per la visita 
	for i in range (0, nNodi+1):
		if(not visitati[i]): 
			sorgente = i
			break
	#Aumenta il numero di nodi visitati e visita il nodo
	contaVisite+=1 
	visita(sorgente)
	#Per ogni arco nello heap, se il nodo di destinazione non è stato visitato, lo visita e aggiunge l'arco alla MSF 
	while(len(heap_archi) > 0):
		a = heapq.heappop(heap_archi) #Estrae ogni volta l'arco di peso minimo
		if(not visitati[a.v]):
			contaVisite+=1
			visita(a.v)
			hash_map[hash((min(a.u,a.v),max(a.u,a.v)))].msf = True
			costoMSF+=a.w
			numCoCo-=1
print(f"{nArchi} {numCoCo} {costoMSF}")

File: Msf.java
import java.io.FileReader;
import java.io.BufferedReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Arrays;

public class Msf{
	public static class Arco implements Comparable<Arco>{
		int u, v, w;
		boolean msf;
		Arco(int u, int v, int w){
			this.u = u;
			this.v = v;
			this.w = w;
			this.msf = false;
		}

		@Override
		public int compareTo(Arco other){
			return Integer.compare(this.w, other.w);
		}
	}

	public static int hash(int u, int v){
		long risultato = (u * 2654435761L) ^ (v * 2246822519L);
		return (int) Math.abs(risultato);
	}
	public static int findParent(int[] parent, int component){ //Find con path compression ricorsiva
		if(parent[component] == component) return component;
		return parent[component] = findParent(parent, parent[component]);
	}
	public static void union(int[] parent, int[] rank, int u, int v){ //Union con union by rank 
		int rootU = findParent(parent, u);
		int rootV = findParent(parent, v);
		if(rootU == rootV) return;
		if(rank[rootU] < rank[rootV]){
			parent[rootU] = rootV;
		}else if(rank[rootU] > rank[rootV]){
			parent[rootV] = rootU;
		}else{
			parent[rootV] = rootU;
			rank[rootU]++;
		}
	}

	public static void main(String[] args){
		if(args.length < 2){
			System.err.println("Utilizzo: java Msf file_grafo file_operazioni");
			System.exit(1);
		}
		String file_grafo = args[0];
		String file_operazioni = args[1];
		final HashMap<Integer, Arco> grafo = new HashMap<>(); //Hashmap degli archi
		ArrayList<Arco> archi = new ArrayList<>(); //Arraylist di archi per kruskal
		int nNodi = -1, nArchi = -1, numCoCo;
/*-------------------------------------------LETTURA GRAFO---------------------------------------------------------------- */
		try(BufferedReader in = new BufferedReader(new FileReader(file_grafo))){
			while(true){
				String op = in.readLine();
				if(op == null) break;
				String[] campi = op.trim().split("\\s+");
				

				switch(campi[0]){
					case "c": //Se trova un commento lo ignora 
						continue;
					case "p": //Se trova la linea di configurazione inizializza i campi
						nNodi = Integer.parseInt(campi[2]) + 1 ;
						nArchi = Integer.parseInt(campi[3]);
						break;
					case "a":
						if(nNodi == -1 || nArchi == -1){
							System.err.println("Linea di configurazione non trovata, file non valido");
							System.exit(1);
						}
						//aggiungi arco
						int u = Integer.parseInt(campi[1]);
						int v = Integer.parseInt(campi[2]);
						if(u>v){
							int tmp = u;
							u = v;
							v = tmp;
						}
						int w = Integer.parseInt(campi[3]);
						Arco a  = new Arco(u, v, w);
						grafo.put(Msf.hash(u, v), a);
						archi.add(a);
						break;
					default:
						continue;
				}
			}
		} catch(Exception e){
			System.err.println("Errore: " + e.getMessage());
			e.printStackTrace(System.err);
			System.exit(1);
		}

/*-----------------------------------------------OPERAZIONI--------------------------------------------------- */
	try(BufferedReader in = new BufferedReader(new FileReader(file_operazioni))){
		while(true){
				String op = in.readLine();
				if(op == null) break;
				String[] campi = op.trim().split("\\s+");
				
				switch(campi[0]){
					case "c": //Se trova un commento lo ignora 
						continue;
					case "+": //Se trova la linea di configurazione inizializza i campi
						int u = Integer.parseInt(campi[1]);
						int v = Integer.parseInt(campi[2]);
						if(u>v){
							int tmp = u;
							u = v;
							v = tmp;
						}
						int w = Integer.parseInt(campi[3]);
						Arco a  = new Arco(u, v, w);
						int key = hash(u, v);
						Arco cmp = grafo.get(key);
						if(cmp != null && cmp.equals((a))){
							System.err.println("Arco da inserire già esistente");
							System.exit(1);
						}
						grafo.put(hash(u, v), a);
						archi.add(a);
						nArchi+=1;
						break;
					case "-":
						u = Integer.parseInt(campi[1]);
						v = Integer.parseInt(campi[2]);
						if(u>v){
							int tmp = u;
							u = v;
							v = tmp;
						}
						Arco rem = grafo.remove(hash(u, v));
						if(rem == null){
							System.err.println("Arco da rimuovere non trovato");
							System.exit(1);
						}
						archi.remove(rem);
						nArchi-=1;
						break;
					default:
						continue;
				}
			}
	} catch(Exception e){
		System.err.println("Errore: " + e.getMessage());
		e.printStackTrace(System.err);
		System.exit(1);
	}
/*-------------------------------------------------------------KRUSKAL---------------------------------------------------------------- */
	archi.sort(Arco::compareTo); //Ordina gli archi per peso crescente
	//Array di parent e rank per l'unione dei componenti con path compression
	int[] parent = new int[nNodi]; 
	for(int i = 0; i < nNodi; i++){
		parent[i] = i;
	}
	int [] rank = new int[nNodi];
	Arrays.fill(rank, 1);

	numCoCo = nNodi;
	long costoMSF = 0;
	for(Arco arco : archi){
		int r1 = findParent(parent, arco.u);
		int r2 = findParent(parent, arco.v);
		int w = arco.w;

		if(r1 != r2){
			union(parent, rank, r1, r2);
			arco.msf = true;
			costoMSF += w;
			numCoCo--;
		}
	}
	System.out.println(nArchi + " " + numCoCo + " " + costoMSF);
}
}